cyences_app_for_splunk/bin/device_inventory_v2_util.py

```diff
@@ -49,349 +114,425 @@ def _internal_check_list_field_format(self, field):
         elif type(field) == list:
             return [element.lower() for element in field]
         else:
-            raise Exception("Unexpected field format. Allowed only str and list of string.")
+            raise Exception(
+                "Unexpected field format. Allowed only str and list of string."
+            )
 
 
-# this is not been in use
-class DeviceField:
-    def __init__(self, value, time, expiry=None) -> None:
-        self.value = value
-        self.time = time
-        self.expiry = expiry
-
-
-class Device:
-    def __init__(self, uuid) -> None:
-        self.uuid = uuid
-
-        self.products = {}   # {<product_name>: {<product_uuid>: {time: <time-epoch>, ips: [<ips>], mac_addresses: [<mac_addresses>], hostnames: [<hostnames>], custom_fields: [<custom-fields>]}}}
-
-        self.product_names = dict()   # {"<name>": <occurrences>}
-        self.product_uuids = dict()   # {"<name>": <occurrences>}
-        self.ips = dict()
-        self.mac_addresses = dict()
-        self.hostnames = dict()
-
-
-    def two_value_combination_match(self, values1, current_list1, values2, current_list2, is_list1_hostname=False, is_list2_hostname=False, hostname_postfixes=[]):
-        # this does comparison in lower case characters always to ensure case-sensitivity does not affect the comparison
-        if is_list1_hostname and hostname_postfixes:
-            values1_updated = [remove_words_from_end(element.lower(), hostname_postfixes) for element in values1]
-            current_list1_updated = [remove_words_from_end(element.lower(), hostname_postfixes) for element in current_list1]
-        else:
-            values1_updated = [element.lower() for element in values1]
-            current_list1_updated = [element.lower() for element in current_list1]
-        
-        if is_list2_hostname and hostname_postfixes:
-            values2_updated = [remove_words_from_end(element.lower(), hostname_postfixes) for element in values2]
-            current_list2_updated = [remove_words_from_end(element.lower(), hostname_postfixes) for element in current_list2]
+class DeviceManager:
+    """
+    Use Example:
+        with DeviceManager() as dm:
+            new_device_entry = DeviceEntry(...)
+            device_uuid = dm.add_device_entry(new_device_entry)
+    """
+
+    def __init__(self, session_key, logger, operation, hostname_postfixes=[]):
+        self.session_key = session_key
+        self.logger = logger
+        self.operation = operation
+        self.hostname_postfixes = [
+            element.strip() for element in hostname_postfixes if element.strip()
+        ]
+        self.modify_devices = []
+        self.devices = self.read_kvstore_lookup(DEVICE_INVENTORY_LOOKUP_COLLECTION)
+        if self.devices:
+            self._convert_str_to_dict()
```

Comment:
I would suggest keeping this inside the read_kvstore_lookup function itself.
OR
use self.read_devices(<lookup>)
            self._read_kvstore_lookup(<lookup>)
            and _convert here

Reply:
like update_kvstore_entries is handling it internally.

Reply:
Keep the self.devices up to date.
* Update self.devices when updated. And add the key to modified_device_list. (While storing build the dict by using self.devices) -> For memory optimization and avoid conflicts
* Remove self.devices when deleted. And add the key to deleted_device_list.

Reply:
Done